File: backend/routers/inventory.py
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from database import get_db
from models import InventoryCheck, InventoryItem, Product, User
from schemas import InventoryCheckCreate, InventoryCheckOut, InventoryItemOut
from auth import get_current_user

router = APIRouter(prefix="/api/inventory", tags=["Inventory"])


def _check_to_out(check: InventoryCheck) -> InventoryCheckOut:
    out = InventoryCheckOut.model_validate(check)
    out.created_by_name = check.created_by_user.full_name if check.created_by_user else None
    items_out = []
    for item in check.items:
        io = InventoryItemOut.model_validate(item)
        io.product_name = item.product.name if item.product else None
        io.product_sku = item.product.sku if item.product else None
        items_out.append(io)
    out.items = items_out
    return out


@router.get("/", response_model=list[InventoryCheckOut])
def list_checks(db: Session = Depends(get_db)):
    checks = db.query(InventoryCheck).order_by(InventoryCheck.created_at.desc()).limit(20).all()
    return [_check_to_out(c) for c in checks]


@router.get("/{check_id}", response_model=InventoryCheckOut)
def get_check(check_id: int, db: Session = Depends(get_db)):
    check = db.query(InventoryCheck).filter(InventoryCheck.id == check_id).first()
    if not check:
        raise HTTPException(status_code=404, detail="Inventarizatsiya topilmadi")
    return _check_to_out(check)


@router.post("/", response_model=InventoryCheckOut, status_code=201)
def create_check(
    data: InventoryCheckCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    check = InventoryCheck(created_by=current_user.id, status="in_progress")
    db.add(check)
    db.flush()

    for item_data in data.items:
        product = db.query(Product).filter(Product.id == item_data.product_id).first()
        if not product:
            continue
        diff = item_data.actual_quantity - product.current_stock
        inv_item = InventoryItem(
            check_id=check.id,
            product_id=item_data.product_id,
            system_quantity=product.current_stock,
            actual_quantity=item_data.actual_quantity,
            difference=diff,
            notes=item_data.notes,
        )
        db.add(inv_item)

    db.commit()
    db.refresh(check)
    return _check_to_out(check)


@router.post("/{check_id}/complete", response_model=InventoryCheckOut)
def complete_check(check_id: int, db: Session = Depends(get_db), _=Depends(get_current_user)):
    check = db.query(InventoryCheck).filter(InventoryCheck.id == check_id).first()
    if not check:
        raise HTTPException(status_code=404, detail="Inventarizatsiya topilmadi")
    if check.status == "completed":
        raise HTTPException(status_code=400, detail="Allaqachon yakunlangan")

    # Apply actual quantities to products
    for item in check.items:
        product = db.query(Product).filter(Product.id == item.product_id).first()
        if product:
            product.current_stock = item.actual_quantity

    check.status = "completed"
    check.completed_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(check)
    return _check_to_out(check)

File: server/routes/inventory.js
import { Router } from 'express';
import { InventoryCheck, InventoryItem, Product } from '../models/index.js';
import sequelize from '../config/database.js';

const router = Router();

// GET /api/inventory — Get all inventory checks
router.get('/', async (req, res) => {
    try {
        const checks = await InventoryCheck.findAll({
            include: [{ model: InventoryItem, as: 'items', include: [{ model: Product, as: 'product' }] }],
            order: [['createdAt', 'DESC']],
        });
        res.json(checks);
    } catch (error) {
        res.status(500).json({ error: 'Inventarizatsiyalarni olishda xatolik' });
    }
});

// POST /api/inventory/start — Start new inventory check
router.post('/start', async (req, res) => {
    const t = await sequelize.transaction();
    try {
        const check = await InventoryCheck.create({
            status: 'in_progress',
            created_by: req.user?.id || 1,
        }, { transaction: t });

        // Create an inventory item for each product
        const products = await Product.findAll({ transaction: t });
        const items = products.map(p => ({
            check_id: check.id,
            product_id: p.id,
            system_quantity: parseFloat(p.current_stock),
            actual_quantity: null,
            difference: 0,
        }));

        await InventoryItem.bulkCreate(items, { transaction: t });
        await t.commit();

        const fullCheck = await InventoryCheck.findByPk(check.id, {
            include: [{ model: InventoryItem, as: 'items', include: [{ model: Product, as: 'product' }] }],
        });
        res.status(201).json(fullCheck);
    } catch (error) {
        await t.rollback();
        console.error('Start inventory error:', error);
        res.status(500).json({ error: 'Inventarizatsiya boshlashda xatolik' });
    }
});

// PUT /api/inventory/:id/items — Update actual quantities
router.put('/:id/items', async (req, res) => {
    try {
        const { items } = req.body; // Array of { id, actual_quantity }

        for (const item of items) {
            const invItem = await InventoryItem.findByPk(item.id);
            if (invItem) {
                const actual = parseFloat(item.actual_quantity || 0);
                const diff = actual - parseFloat(invItem.system_quantity);
                await invItem.update({
                    actual_quantity: actual,
                    difference: diff,
                    notes: item.notes || null,
                });
            }
        }

        const check = await InventoryCheck.findByPk(req.params.id, {
            include: [{ model: InventoryItem, as: 'items', include: [{ model: Product, as: 'product' }] }],
        });
        res.json(check);
    } catch (error) {
        console.error('Update inventory items error:', error);
        res.status(500).json({ error: 'Inventarizatsiya ma\'lumotlarini yangilashda xatolik' });
    }
});

// POST /api/inventory/:id/complete — Finalize inventory: update product stocks
router.post('/:id/complete', async (req, res) => {
    const t = await sequelize.transaction();
    try {
        const check = await InventoryCheck.findByPk(req.params.id, {
            include: [{ model: InventoryItem, as: 'items' }],
            transaction: t,
        });

        if (!check) {
            await t.rollback();
            return res.status(404).json({ error: 'Inventarizatsiya topilmadi' });
        }

        if (check.status === 'completed') {
            await t.rollback();
            return res.status(400).json({ error: 'Bu inventarizatsiya allaqachon yakunlangan' });
        }

        // Update product stocks based on actual quantities
        for (const item of check.items) {
            if (item.actual_quantity !== null) {
                await Product.update(
                    { current_stock: parseFloat(item.actual_quantity) },
                    { where: { id: item.product_id }, transaction: t }
                );
            }
        }

        await check.update({
            status: 'completed',
            completed_at: new Date(),
        }, { transaction: t });

        await t.commit();

        const fullCheck = await InventoryCheck.findByPk(check.id, {
            include: [{ model: InventoryItem, as: 'items', include: [{ model: Product, as: 'product' }] }],
        });
        res.json(fullCheck);
    } catch (error) {
        await t.rollback();
        console.error('Complete inventory error:', error);
        res.status(500).json({ error: 'Inventarizatsiyani yakunlashda xatolik' });
    }
});

export default router;
